rustcBeta: Fix build

the-kenny opened the pull request:
###### Motivation for this change

This PR updates `rustcBeta.rustc` and `rustcBeta.cargo` to their newest versions. It also fixes the builds which are broken since the update to `rustc-1.10.0`.

This is a partial fix for #17097 as it updates `rustUnstable` which now fails its own test suite.
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [x] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

the-kenny committed:
rustBeta.rustc: beta-1.10.10 -> 1.11.0, rustBeta.cargo: 0.10.0 -> 0.12.0.

the-kenny committed:
rustBeta.cargo: Fix build.

the-kenny committed:
rustUnstable.rustc: master-1.11.0 -> master-1.12.0

Note that tests are failing for this commit.

garbas commented:
@the-kenny should we wait until you find time for that broken test for rustUnstable or is this already an improvement over existing state?

matthewbauer commented:
"rustc-bootstrap" depends on patchelf which is unavailable on Darwin. Can we disable that platform?

the-kenny commented:
@garbas it's definitely an improvement as `beta` now works. We need to check the darwin issue @matthewbauer mentioned and then I'd say :+1: for merge.

the-kenny commented:
@matthewbauer Can we just skip the `patchelf` calls on darwin? Could you try commenting them out and see if everything builds for you?

the-kenny commented:
Regarding `unstable`: I reported the `tcp-stress-test.rs` failure to upstream - let's see what comes out of it: https://github.com/rust-lang/rust/issues/35107

the-kenny committed:
rust-bootstrap: Don't run patchelf on darwin.

the-kenny commented:
Testing it on darwin now.

the-kenny commented:
Stable on darwin fails massively in the test suite :( No idea where that comes from:

```
failures:
    sys::ext::net::test::basic
    sys::ext::net::test::iter
    sys::ext::net::test::test_connect_unix_datagram
    sys::ext::net::test::test_read_timeout
    sys::ext::net::test::test_read_with_timeout
    sys::ext::net::test::test_unix_datagram
    sys::ext::net::test::test_unix_datagram_recv
    sys::ext::net::test::test_unnamed_unix_datagram
    sys::ext::net::test::timeouts
    sys::ext::net::test::try_clone

test result: FAILED. 739 passed; 10 failed; 2 ignored; 0 measured
```

This might be related to the very horribly broken and misconfigured system I'm using for building here - I don't have access to a clean machine. 

Anyway - the bootstrap code now works fine. Fixing the build itself should be a different issue (and this PR is about `unstable` and `beta` anyway)

garbas commented:
/me gives a try on a clean osx box

the-kenny committed:
rustUnstable.rustc: Fix tests by disabling lowering thread-count.

the-kenny commented:
^ Above builds fine on my NixOS instance. Please test the triage (`rustc`, `rustBeta.rustc`, `rustUnstable.rustc`) on your machines and especially on Darwin.

@garbas: Any results?

globin commented:
On our build machine:
```
failures:
    sys::ext::net::test::basic
    sys::ext::net::test::iter
    sys::ext::net::test::test_connect_unix_datagram
    sys::ext::net::test::test_read_timeout
    sys::ext::net::test::test_read_with_timeout
    sys::ext::net::test::test_unix_datagram
    sys::ext::net::test::test_unix_datagram_recv
    sys::ext::net::test::test_unnamed_unix_datagram
    sys::ext::net::test::timeouts
    sys::ext::net::test::try_clone

test result: FAILED. 739 passed; 10 failed; 2 ignored; 0 measured

make: *** [/private/var/folders/d4/0bw6hrcd1s5962bw7k_k7mjc0000gn/T/nix-build-rustc-1.10.0.drv-0/rust-cfcb716/mk/tests.mk:423: tmp/check-stage2-T-x86_64-apple-darwin-H-x86_64-apple-darwin-std.ok] Error 101
builder for ‘/nix/store/sabz2266nj3gkmm63zksbnjwi4qnsn5v-rustc-1.10.0.drv’ failed with exit code 2
error: build of ‘/nix/store/sabz2266nj3gkmm63zksbnjwi4qnsn5v-rustc-1.10.0.drv’ failed
```

matthewbauer commented:
Okay, can doCheck be set to !stdenv.isDarwin? I'm thinking it's just an upstream issue.

the-kenny committed:
rustcUnstable: Add note about when to drop the patch.

the-kenny committed:
rustc: Disable doCheck on Darwin.

the-kenny commented:
Ok, please retry :)

garbas commented:
@the-kenny for some reason tmux session is started hanged, i'm rerunning with you latest changes

globin commented:
Fails on `rustBeta.rustc`:
```
make[1]: *** No rule to make target 'clang_rt.builtins_x86_64_osx'.  Stop.
make[1]: Leaving directory '/private/var/folders/d4/0bw6hrcd1s5962bw7k_k7mjc0000gn/T/nix-build-rustc-beta-1.11.0-g9333c42.drv-0/rust-9333c42/x86_64-apple-darwin/rt/compiler-rt'
make: *** [/private/var/folders/d4/0bw6hrcd1s5962bw7k_k7mjc0000gn/T/nix-build-rustc-beta-1.11.0-g9333c42.drv-0/rust-9333c42/mk/rt.mk:469: x86_64-apple-darwin/rt/libcompiler-rt.a] Error 2
builder for ‘/nix/store/lrayrdvb6nr43spwpf45i115dia5lcy5-rustc-beta-1.11.0-g9333c42.drv’ failed with exit code 2
error: build of ‘/nix/store/lrayrdvb6nr43spwpf45i115dia5lcy5-rustc-beta-1.11.0-g9333c42.drv’ failed
```

garbas commented:
darwin: ``nix-build -A rustStable`` worked for me
darwin: ``nix-build -A rustBeta`` failed with error that @globin repoted

building rustUnstable right now.

@the-kenny why does it try to build rust-beta when i trigger ``nix-build -A rustUnstable``? didn't check the expressions, just asking if this is expected.

the-kenny commented:
@garbas Expected - rustUnstable is bootstrapped from rustBeta, and rustBeta is bootstrapped from          rustStable. That's how rust bootstraps itself from now on.

`rustStable` works on my OS X box now - but cargo with segmentation fault when fetching the registry. Can someone please try it?

garbas commented:
@the-kenny i can confirm that

``` shell
cltbld@t-yosemite-r7-0003.test.releng.scl3.mozilla.com tmptmp]$ ./../tmp/bin/cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
Segmentation fault: 11
```

garbas commented:
@the-kenny it would be nice to have some doInstallCheck tests to check if basic stuff works with rust.

the-kenny commented:
@garbas Incoming.

I wonder what's the issue with cargo is though :/ Quite sure that's out of my league to fix.

the-kenny commented:
@garbas On second thought: What is there to check in `doInstallCheck`? Just testing if `rustc` and `cargo` are callable?

the-kenny commented:
Pulled a backtrace for that `segmentation fault 11` out of gdb. Maybe someone can make sense out of it:

```
#0  0x000000010076aac8 in CFRelease () from /nix/store/7akpgkhmbmdc9nj69wv0wr01pp9z92x6-CF-osx-10.9.5/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
#1  0x000000010038e31b in stransport_connect ()
#2  0x00000001003979de in http_connect ()
#3  0x0000000100396f46 in http_action ()
#4  0x000000010039a184 in git_smart.connect ()
#5  0x0000000100383833 in git_remote_connect ()
#6  0x0000000100384002 in git_remote_fetch ()
#7  0x0000000100329504 in git2::remote::Remote::fetch::hf22a703e28682659 ()
#8  0x000000010021bda7 in cargo::sources::git::utils::fetch::_$u7b$$u7b$closure$u7d$$u7d$::h4505ef50fa16b942 ()
#9  0x0000000100214bab in cargo::sources::git::utils::fetch::h969cb9309bcaae70 ()
#10 0x0000000100237a2c in cargo::sources::registry::RegistrySource::do_update::hf310bfc89b20f31c ()
#11 0x000000010008ee5b in _$LT$sources..registry..RegistrySource$LT$$u27$cfg$GT$$u20$as$u20$core..source..Source$GT$::update::h2ca6fb87f187829e ()
#12 0x00000001000de130 in cargo::core::registry::PackageRegistry::ensure_loaded::h756a0c74cb30f33c ()
#13 0x00000001000e4d6b in _$LT$core..registry..PackageRegistry$LT$$u27$cfg$GT$$u20$as$u20$core..registry..Registry$GT$::query::h2b3b59a668f8397b ()
#14 0x00000001000cd272 in _$LT$$RF$$u27$a$u20$mut$u20$I$u20$as$u20$std..iter..Iterator$GT$::next::h2ee138e44a29d8c2 ()
#15 0x00000001000c6811 in cargo::core::resolver::Context::build_deps::h5d514997c1c12bc9 ()
#16 0x00000001000bedc4 in cargo::core::resolver::activate::h84824bf29ba071ae ()
#17 0x00000001000b8b02 in cargo::core::resolver::activate_deps_loop::he29eb0ad00974f9d ()
#18 0x00000001000b7004 in cargo::core::resolver::resolve::h8a35d33c5144340a ()
#19 0x00000001000f98fe in cargo::ops::resolve::resolve_with_previous::h4a25e46674f9dd8c ()
#20 0x00000001000f6cca in cargo::ops::resolve::resolve_pkg::h02c4f36849c2231c ()
#21 0x00000001000f50f9 in cargo::ops::cargo_compile::resolve_dependencies::hcb3c885ff9b99a6f ()
#22 0x00000001000f1c94 in cargo::ops::cargo_compile::compile_pkg::hb5fc06c82c8470e8 ()
#23 0x00000001000f0a8e in cargo::ops::cargo_compile::compile::h8a73cb42c0d5b921 ()
#24 0x00000001000176a6 in cargo::call_main_without_stdin::h47b6c412d41ff219 ()
#25 0x00000001000087e1 in cargo::execute::h8bbdddbb8cbc7a98 ()
#26 0x0000000100004ea4 in cargo::call_main_without_stdin::h25a1baa5cba46258 ()
#27 0x0000000100001db1 in cargo::main::h2b219a79f378c1ef ()
#28 0x0000000100422963 in std::sys_common::unwind::try::try_fn::h09ba69fd13531e58 ()
#29 0x000000010041a09c in __rust_try ()
#30 0x00000001004227a5 in std::rt::lang_start::h5b0863080165c75e ()
#31 0x00000001000018c4 in start ()
```

the-kenny commented:
@garbas I'm not sure if `installCheck`s would help - cargo and rustc work fine, cargo only fails when fetching the registry.

the-kenny commented:
the backtrace stems somewhere from libgit2 and is very strange :/ I'm still trying to track down the original issue. 

As a workaround I'm trying to disable `Security.framework` in `libgit2` on Darwin right now. If that works, I'd like to add this to the PR and merge it to prevent bit-rot.

LnL7 committed:
rustc: Set TMPDIR on Darwin to fix build.

the-kenny committed:
libgit2: Disable Security.framework transport on Darwin.

When used by cargo, `libgit2` crashes with a Segmentation Fault on
Darwin. This crash is somehow connected to the `Security.framework`
native to Darwin, and while being easy to reproduce, is very hard to
track down.

This commit introduces a patch to `libgit2` which disables the
troublesome transport and instead depends on `libcurl`. The patch also
adds support for `SSL_CERT_FILE` to `libgit2`.

Upstream tracking issue is
https://github.com/libgit2/libgit2/issues/3885.

the-kenny committed:
cargo: Force usage of pkgs.cacert.

the-kenny committed:
rustBeta.cargo: cargo -> 2016-07-25.

the-kenny committed:
rustHead.cargo: cargo -> 2016-07-25.

the-kenny commented:
Ok! These latest changes (before the rebase against master) finally contain a workaround for that dreadful segmentation fault on Darwin.

I wasn't able to track down the issue completely yet, and the workaround isn't the nicest, but at least it's working fine.

Can you all please test on NixOS and Darwin? I tested `rustStable` and `rustBeta` on NixOS and `rustStable` on Darwin. `rustUnstable` is building on my machine right now.

LnL7 commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -107,6 +115,7 @@ stdenv.mkDerivation {
 
-  doCheck = true;
+  # Disable doCheck on Darwin to work around upstream issue
+  doCheck = !stdenv.isDarwin;
```
checkout #17584, if you set TMPDIR the tests will pass on darwin.

LnL7 commented on pkgs/development/compilers/rust/bootstrap.nix:
```diff
@@ -46,9 +48,11 @@ rec {
-      patchelf \
-        --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
-        "$out/bin/rustc"
+      ${if needsPatchelf then ''
```
you can use `stdenv.lib.optionalString` here

the-kenny replied:
Thanks!

the-kenny committed:
rustc: Re-enable doCheck on Darwin.

the-kenny committed:
rustc-bootstrap: Use optionalString.

the-kenny commented:
Okay, evaluation doesn't work on Darwin as `http-parser` is marked on darwin as of ea14f5c86dd1e28e4cf367c311353efb713e1a9e

Any ideas? `http-parser` is a dependency of `libgit2`

the-kenny commented:
Created https://github.com/NixOS/nixpkgs/pull/17609 to gather feedback on reverting the change that disables evaluation of `rust` and `node` on darwin.

garbas commented:
@the-kenny should we just go ahead and merge this PR as it is improvement over current state and we focus on rust+darwin in future PRs. just so this work doesn't get stale.

the-kenny commented:
@garbas :+1: I think in general this PR still improves the status quo

LnL7 commented:
@the-kenny did you test this on a linux platform? https://gist.github.com/LnL7/0777081c6373fd7e2cd27b77f291023f

the-kenny commented:
@LnL7 No :/ I don't have access to a linux platform at the moment. Need to spin up a VM or so.

the-kenny commented:
@LnL7 I suppose we can just disable the test for now. This non-utf8-linker-output test looks like a strange corner case to me ;)

the-kenny committed:
rustc: Disable failing linker-output-utf8 test.

LnL7 commented:
I agree, this might even work depending on what filesystem you use. Just disable it and we can merge this in.

the-kenny commented:
@LnL7 Above patch disables the utf8 test. Can you please try again on your linux machine?

the-kenny commented:
Rebasing against master now.

LnL7 commented:
@the-kenny There's no easier way to disable a test? Seems like a pretty big patch

the-kenny commented:
It just deletes the files. Let me check if there's an easier way (that also works on all three compilers)

the-kenny commented:
We could just `rm -r` the directory. That may be cleaner.

the-kenny commented:
@LnL7 Force-pushed a nicer way of disabling the test :)
